pkg/urbit/include/c/defs.h

```diff
@@ -157,4 +159,108 @@
+      {
+        static const size_t max_attempts_i    = 100;
+        size_t              attempts_i        = 0;
+        c3_y*               ptr_y             = data_v;
+        ssize_t             bytes_remaining_i = data_len_i;
+        do {
+          attempts_i++;
+          ssize_t bytes_read_i = read(fd_i, ptr_y, bytes_remaining_i);
+          if ( -1 == bytes_read_i ) {
+            if ( (EINTR == errno || EAGAIN == errno || EWOULDBLOCK == errno)
+                 && attempts_i <= max_attempts_i )
+            {
+              continue;
+            }
+            else {
+              fprintf(stderr,
+                      "read: failed after %lu attempts: %s\r\n",
+                      attempts_i,
+                      strerror(errno));
+              return 0;
+            }
+          }
+          else if ( 0 == bytes_read_i ) {
+            break;
+          }
+          else {
+            ptr_y += bytes_read_i;
+            bytes_remaining_i -= bytes_read_i;
+            attempts_i = 0;
+          }
+        } while ( bytes_remaining_i > 0 );
+        return 1;
+      }
+
+      /// Writes the contents of a buffer to a file.
+      ///
+      /// This function correctly handles the case in which write() writes fewer
+      /// bytes than was requested by retrying until all bytes have been written
+      /// or until an error occurs, whichever comes first.
+      ///
+      /// If interrupted by a signal or if the write() call would block, this
+      /// function will retry up to a specified maximum number of attempts.
+      ///
+      ///
+      /// @param[in] fd_i        Open file descriptor to write to.
+      /// @param[in] data_v      Source buffer.
+      /// @param[in] data_len_i  Length of `data_v` in bytes.
+      ///
+      /// @return 1  Complete write succeeded.
+      /// @return 0  Otherwise.
+      static inline c3_t
+      c3_write(c3_i fd_i, const void* const data_v, const size_t data_len_i)
+      {
+        static const size_t max_attempts_i    = 100;
+        size_t              attempts_i        = 0;
+        const c3_y*         ptr_y             = data_v;
+        ssize_t             bytes_remaining_i = data_len_i;
+        do {
+          attempts_i++;
+          ssize_t bytes_written_i = write(fd_i, ptr_y, bytes_remaining_i);
+          if ( -1 == bytes_written_i ) {
+            // Attempt to read again if we were interrupted by a signal.
+            if ( (EINTR == errno || EAGAIN == errno || EWOULDBLOCK == errno)
+                 && attempts_i <= max_attempts_i )
+            {
+              continue;
+            }
+            else {
+              fprintf(stderr,
+                      "write: failed after %lu attempts: %s\r\n",
+                      attempts_i,
+                      strerror(errno));
+              return 0;
+            }
+          }
+          else if ( 0 == bytes_written_i ) {
+            break;
+          }
```

Comment:
This seems wrong, I think it should just be `continue;` (but could also be removed in favor of the else block below).

Reply:
You're right. I've removed this `else if`.